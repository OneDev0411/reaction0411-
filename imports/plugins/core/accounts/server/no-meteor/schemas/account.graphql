"The fields by which you are allowed to sort any query that returns an `AccountConnection`"
enum AccountSortByField {
  "Account ID"
  _id

  "Date and time at which the account was created"
  createdAt

  "Account name"
  name

  "Date and time at which the account was last updated"
  updatedAt
}

"Defines a new Address and the account to which it should be added"
input AddAccountAddressBookEntryInput {
  "The account ID"
  accountId: ID!

  "The address to add"
  address: AddressInput!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String
}

"Describes changes that should be applied to one of the addresses for an account"
input UpdateAccountAddressBookEntryInput {
  "The account ID"
  accountId: ID!

  "The address ID"
  addressId: ID!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "If present, make this address the default address of this type"
  type: AddressType

  "The address changes to apply"
  updates: AddressInput!
}

"Describes which address should be removed from which account"
input RemoveAccountAddressBookEntryInput {
  "The account ID"
  accountId: ID!

  "The address ID"
  addressId: ID!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String
}

"Describes which email should be sent a password reset link"
input ResetAccountPasswordInput {
  "The email address of the account to send reset email to"
  email: String!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String
}

"Describes an account profile currency change"
input SetAccountProfileCurrencyInput {
  "The account ID, which defaults to the viewer account"
  accountId: ID

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The currency code"
  currencyCode: String!
}

"Defines a new Email and the account to which it should be added"
input AddAccountEmailRecordInput {
  "The account ID, which defaults to the viewer account"
  accountId: ID

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The email address to add"
  email: Email!
}

"Defines which email address should be removed from which account"
input RemoveAccountEmailRecordInput {
  "The account ID, which defaults to the viewer account"
  accountId: ID

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The email address to remove"
  email: Email!
}

"Per-account tax exemption settings used by the Avalara plugin"
type TaxSettings {
  "Customer usage type. A value matching the `code` field of one TaxEntityCode, or any custom string."
  customerUsageType: String

  "Exemption number for an account"
  exemptionNo: String
}

"Represents a single user account"
type Account implements Node {
  "The account ID"
  _id: ID!

  "A list of physical or mailing addresses associated with this account"
  addressBook(
    "Return only results that come after this cursor. Use this with `first` to specify the number of results to return."
    after: ConnectionCursor,

    "Return only results that come before this cursor. Use this with `last` to specify the number of results to return."
    before: ConnectionCursor,

    "Return at most this many results. This parameter may be used with either `after` or `offset` parameters."
    first: ConnectionLimitInt,

    "Return at most this many results. This parameter may be used with the `before` parameter."
    last: ConnectionLimitInt
  ): AddressConnection

  "The date and time at which this account was created"
  createdAt: DateTime!

  "The preferred currency used by this account"
  currency: Currency

  "A list of email records associated with this account"
  emailRecords: [EmailRecord]

  "The first name of the person this account represents, if known"
  firstName: String

  "A paged list of the permission groups in which this account is listed"
  groups(
    "Return only results that come after this cursor. Use this with `first` to specify the number of results to return."
    after: ConnectionCursor,

    "Return only results that come before this cursor. Use this with `last` to specify the number of results to return."
    before: ConnectionCursor,

    "Return at most this many results. This parameter may be used with either `after` or `offset` parameters."
    first: ConnectionLimitInt,

    "Return at most this many results. This parameter may be used with the `before` parameter."
    last: ConnectionLimitInt,

    "Return results sorted in this order"
    sortOrder: SortOrder = asc,

    "By default, groups are sorted by when they were created, oldest first. Set this to sort by one of the other allowed fields"
    sortBy: GroupSortByField = createdAt
  ): GroupConnection

  "The last name of the person this account represents, if known"
  lastName: String

  "Arbitrary additional metadata about this account"
  metafields: [Metafield]

  "The full name of the person this account represents, if known"
  name: String

  "Some note about this account"
  note: String

  "An object storing plugin-specific preferences for this account"
  preferences: JSONObject

  """
  The primary email address for the account. This matches the address in `emailRecords` where `provides` is "default".
  """
  primaryEmailAddress: Email!

  "The shop to which this account belongs, if it is associated with a specific shop"
  shop: Shop

  "Per-account tax exemption settings used by the Avalara plugin"
  taxSettings: TaxSettings

  "The date and time at which this account was last updated"
  updatedAt: DateTime!
}

"""
Wraps a list of `Accounts`, providing pagination cursors and information.

For information about what Relay-compatible connections are and how to use them, see the following articles:
- [Relay Connection Documentation](https://facebook.github.io/relay/docs/en/graphql-server-specification.html#connections)
- [Relay Connection Specification](https://facebook.github.io/relay/graphql/connections.htm)
- [Using Relay-style Connections With Apollo Client](https://www.apollographql.com/docs/react/recipes/pagination.html)
"""
type AccountConnection {
  "The list of nodes that match the query, wrapped in an edge to provide a cursor string for each"
  edges: [AccountEdge]

  """
  You can request the `nodes` directly to avoid the extra wrapping that `NodeEdge` has,
  if you know you will not need to paginate the results.
  """
  nodes: [Account]

  "Information to help a client request the next or previous page"
  pageInfo: PageInfo!

  "The total number of nodes that match your query"
  totalCount: Int!
}

"A connection edge in which each node is an `Account` object"
type AccountEdge implements NodeEdge {
  "The cursor that represents this node in the paginated results"
  cursor: ConnectionCursor!

  "The account"
  node: Account
}

"The response from the `addAccountAddressBookEntry` mutation"
type AddAccountAddressBookEntryPayload {
  "The added address"
  address: Address

  "The added address edge"
  addressEdge: AddressEdge

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

"The response from the `addAccountEmailRecord` mutation"
type AddAccountEmailRecordPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The added email record"
  emailRecord: EmailRecord
}

"The response from the `resetAccountPassword` mutation"
type ResetAccountPasswordPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The email address of the account to send reset email to"
  email: String!
}

"The response from the `removeAccountAddressBookEntry` mutation"
type RemoveAccountAddressBookEntryPayload {
  "The removed address"
  address: Address

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

"The response from the `removeAccountEmailRecord` mutation"
type RemoveAccountEmailRecordPayload {
  "The account, with updated `emailRecords`"
  account: Account

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

"The response from the `updateAccountAddressBookEntry` mutation"
type UpdateAccountAddressBookEntryPayload {
  "The updated address"
  address: Address

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

"The response from the `setAccountProfileCurrency` mutation"
type SetAccountProfileCurrencyPayload {
  "The updated account"
  account: Account

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

extend type Shop {
  """
  Returns a list of administrators for this shop, as a Relay-compatible connection.
  "Administrators" means all linked accounts that have the "admin" role for this shop.
  """
  administrators(
    "Return only results that come after this cursor. Use this with `first` to specify the number of results to return."
    after: ConnectionCursor,

    "Return only results that come before this cursor. Use this with `last` to specify the number of results to return."
    before: ConnectionCursor,

    "Return at most this many results. This parameter may be used with either `after` or `offset` parameters."
    first: ConnectionLimitInt,

    "Return at most this many results. This parameter may be used with the `before` parameter."
    last: ConnectionLimitInt,

    "Return results sorted in this order"
    sortOrder: SortOrder = asc,

    "By default, accounts are sorted by when they were created, oldest first. Set this to sort by one of the other allowed fields"
    sortBy: AccountSortByField = createdAt
  ): AccountConnection
}

extend type Mutation {
  "Add a new address to the `addressBook` field for an account"
  addAccountAddressBookEntry(
    "Mutation input"
    input: AddAccountAddressBookEntryInput!
  ): AddAccountAddressBookEntryPayload

  "Add an email address to an account"
  addAccountEmailRecord(
    "Mutation input"
    input: AddAccountEmailRecordInput!
  ): AddAccountEmailRecordPayload

  "Remove an address from the `addressBook` field for an account"
  removeAccountAddressBookEntry(
    "Mutation input"
    input: RemoveAccountAddressBookEntryInput!
  ): RemoveAccountAddressBookEntryPayload

  "Remove an email address from an account"
  removeAccountEmailRecord(
    "Mutation input"
    input: RemoveAccountEmailRecordInput!
  ): RemoveAccountEmailRecordPayload

  "Send a reset password email to an email address from an account"
  resetAccountPassword(
    "Mutation input"
    input: ResetAccountPasswordInput!
  ): ResetAccountPasswordPayload

  "Set the preferred currency for an account"
  setAccountProfileCurrency(
    "Mutation input"
    input: SetAccountProfileCurrencyInput!
  ): SetAccountProfileCurrencyPayload

  "Remove an address that exists in the `addressBook` field for an account"
  updateAccountAddressBookEntry(
    "Mutation input"
    input: UpdateAccountAddressBookEntryInput!
  ): UpdateAccountAddressBookEntryPayload
}

extend type Query {
  "Returns the account with the provided ID"
  account(
    "The account ID"
    id: ID!
  ): Account

  """
  Returns a list of administrators for the shop with ID `shopId`, as a Relay-compatible connection.
  "Administrators" means all linked accounts that have the "admin" role for this shop.
  """
  administrators(
    "Return accounts that are administrators in this shop"
    shopId: ID!,

    "Return only results that come after this cursor. Use this with `first` to specify the number of results to return."
    after: ConnectionCursor,

    "Return only results that come before this cursor. Use this with `last` to specify the number of results to return."
    before: ConnectionCursor,

    "Return at most this many results. This parameter may be used with either `after` or `offset` parameters."
    first: ConnectionLimitInt,

    "Return at most this many results. This parameter may be used with the `before` parameter."
    last: ConnectionLimitInt,

    "Return results sorted in this order"
    sortOrder: SortOrder = asc,

    "By default, accounts are sorted by when they were created, oldest first. Set this to sort by one of the other allowed fields"
    sortBy: AccountSortByField = createdAt
  ): AccountConnection

  "Returns the account for the authenticated user"
  viewer: Account
}
