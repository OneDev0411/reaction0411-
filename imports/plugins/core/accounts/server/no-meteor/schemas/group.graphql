"The fields by which you are allowed to sort any query that returns an `GroupConnection`"
enum GroupSortByField {
  "Group ID"
  _id

  "Date and time at which this group was created"
  createdAt

  "Group name"
  name

  "Date and time at which this group was last updated"
  updatedAt
}

"A group definition"
input GroupInput {
  "A free text description of this group"
  description: String

  "A unique name for the group"
  name: String!

  "A list of the account permissions implied by membership in this group"
  permissions: [String]

  "A unique URL-safe string representing this group"
  slug: String!
}

"The details for creating a group"
input CreateGroupInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The group to create"
  group: GroupInput!

  "The ID of the shop this group belongs to"
  shopId: ID!
}

"The details for updating a group"
input UpdateGroupInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The group ID"
  id: ID!

  "The ID of the shop this group belongs to"
  shopId: ID!

  "The changes to apply to the group"
  updates: GroupInput!
}

"The details for removing a group"
input RemoveGroupInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The group ID"
  id: ID!
}

"Defines a group and account that should be linked"
input AddAccountToGroupInput {
  "The account ID"
  accountId: ID!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The group ID"
  groupId: ID!
}

"Defines a group and account that should be unlinked"
input RemoveAccountFromGroupInput {
  "The account ID"
  accountId: ID!

  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The group ID"
  groupId: ID!
}

"Represents an account group"
type Group implements Node {
  "The group ID"
  _id: ID!

  "The date and time at which this group was created"
  createdAt: DateTime!

  "The account that created this group"
  createdBy: Account

  "A free text description of this group"
  description: String

  "A unique name for the group"
  name: String!

  "A list of the account permissions implied by membership in this group"
  permissions: [String]

  "The shop to which this group belongs"
  shop: Shop

  "A unique URL-safe string representing this group"
  slug: String!

  "The date and time at which this group was last updated"
  updatedAt: DateTime!
}

"""
Wraps a list of `Groups`, providing pagination cursors and information.

For information about what Relay-compatible connections are and how to use them, see the following articles:
- [Relay Connection Documentation](https://facebook.github.io/relay/docs/en/graphql-server-specification.html#connections)
- [Relay Connection Specification](https://facebook.github.io/relay/graphql/connections.htm)
- [Using Relay-style Connections With Apollo Client](https://www.apollographql.com/docs/react/recipes/pagination.html)
"""
type GroupConnection {
  "The list of nodes that match the query, wrapped in an edge to provide a cursor string for each"
  edges: [GroupEdge]

  """
  You can request the `nodes` directly to avoid the extra wrapping that `NodeEdge` has,
  if you know you will not need to paginate the results.
  """
  nodes: [Group]

  "Information to help a client request the next or previous page"
  pageInfo: PageInfo!

  "The total number of nodes that match your query"
  totalCount: Int!
}

"A connection edge in which each node is a `Group` object"
type GroupEdge implements NodeEdge {
  "The cursor that represents this node in the paginated results"
  cursor: ConnectionCursor!

  "The group"
  node: Group
}

"The response from the `addAccountToGroup` mutation"
type AddAccountToGroupPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The updated group"
  group: Group
}

"The response from the `createGroup` mutation"
type CreateGroupPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The new group"
  group: Group
}

"The response from the `updateGroup` mutation"
type UpdateGroupPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The updated group"
  group: Group
}

"The response from the `removeAccountFromGroup` mutation"
type RemoveAccountFromGroupPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The updated group"
  group: Group
}

"The response from the `removeGroup` mutation"
type RemoveGroupPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "Successfully removed?"
  wasRemoved: Boolean!
}

extend type Mutation {
  "Add an account to a group"
  addAccountToGroup(input: AddAccountToGroupInput!): AddAccountToGroupPayload

  "Create a new permission group"
  createGroup(input: CreateGroupInput!): CreateGroupPayload

  "Remove an account from a group"
  removeAccountFromGroup(input: RemoveAccountFromGroupInput!): RemoveAccountFromGroupPayload

  "Remove an existing permission group"
  removeGroup(input: RemoveGroupInput!): RemoveGroupPayload

  "Update an existing permission group"
  updateGroup(input: UpdateGroupInput!): UpdateGroupPayload
}

extend type Query {
  "Returns a single group by ID."
  group(id: ID!): Group

  "Returns a list of groups for the shop with ID `shopId`, as a Relay-compatible connection."
  groups(shopId: ID!, after: ConnectionCursor, before: ConnectionCursor, first: ConnectionLimitInt, last: ConnectionLimitInt, sortOrder: SortOrder = asc, sortBy: GroupSortByField = createdAt): GroupConnection
}
