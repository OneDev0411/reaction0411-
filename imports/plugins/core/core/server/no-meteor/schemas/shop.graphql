"Input parameters for the inviteShopMember mutation"
input InviteShopMemberInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "The email address of the person to invite"
  email: String!

  "The permission group for this person's new account"
  groupId: ID!

  "The invitee's full name"
  name: String!

  "The ID of the shop to which you want to invite this person"
  shopId: ID!
}

"The response from the `inviteShopMember` mutation"
type InviteShopMemberPayload {
  "The account that was successfully created or found and updated by inviting this shop member"
  account: Account

  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String
}

extend type Mutation {
  """
  Given a person's email address and name, invite them to create an account for a certain shop,
  and put them in the provided permission group
  """
  inviteShopMember(input: InviteShopMemberInput!): InviteShopMemberPayload
}
