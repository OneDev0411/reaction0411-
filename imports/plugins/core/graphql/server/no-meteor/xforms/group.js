import namespaces from "@reactioncommerce/api-utils/graphql/namespaces.js";
import { assocInternalId, assocOpaqueId, decodeOpaqueIdForNamespace, encodeOpaqueId } from "./id";

export const assocGroupInternalId = assocInternalId(namespaces.Group);
export const assocGroupOpaqueId = assocOpaqueId(namespaces.Group);
export const decodeGroupOpaqueId = decodeOpaqueIdForNamespace(namespaces.Group);
export const encodeGroupOpaqueId = encodeOpaqueId(namespaces.Group);
