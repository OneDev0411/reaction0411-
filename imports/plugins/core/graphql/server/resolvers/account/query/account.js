import { transformIdFromBase64 } from "@reactioncommerce/reaction-graphql-utils";
import { userAccountQuery } from "/imports/accounts/server/methods/userAccountQuery";

export default function account(_, { id }) {
  // search for user from the Accounts collection via provided Account ID

/**
 * @name account
 * @method
 * @summary query the Accounts collection and return user account data
 * @param {Object} _ - an object containing the result returned from the resolver
 * @param {Object} args - an object containing Meteor context
 * @param {String} args.id - id of user to query
 * @param {Object} context - an object containing the per-request state
 * @return {Object} user account object
 */
export default function account(_, { id }, context) {
  // Trasform ID from base64
  // Returns an object. Use `.id` to get ID
  const idFromBase64 = transformIdFromBase64(id);

  // Pass Id into userAccountQuery function
  const userAccount = userAccountQuery(idFromBase64.id);

  // Return result of userAccountQuery()
  return userAccount;
}
