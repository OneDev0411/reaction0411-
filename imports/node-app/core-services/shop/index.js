import mutations from "./mutations/index.js";
import queries from "./queries/index.js";
import resolvers from "./resolvers/index.js";
import schemas from "./schemas/index.js";
import createDataLoaders from "./utils/createDataLoaders.js";
import { Shop } from "./simpleSchemas.js";

/**
 * @summary Import and call this function to add this plugin to your API.
 * @param {ReactionNodeApp} app The ReactionNodeApp instance
 * @returns {undefined}
 */
export default async function register(app) {
  await app.registerPlugin({
    label: "Shop",
    name: "reaction-shop",
    graphQL: {
      resolvers,
      schemas
    },
    queries,
    mutations,
    functionsByType: {
      createDataLoaders: [createDataLoaders]
    },
    simpleSchemas: {
      Shop
    }
  });
}
