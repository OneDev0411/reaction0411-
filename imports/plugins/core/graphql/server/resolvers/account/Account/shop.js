import { xformShopResponse } from "@reactioncommerce/reaction-graphql-xforms/shop";

/**
 * @name shop
 * @method
 * @summary gets the shop object for the provided account
 * @param {Object} account - result of the parent resolver, which is an Account object in GraphQL schema format
 * @return {Object} The shop having ID account.shopId, in GraphQL schema format
 */
export default function shop(account) {
  const { shopId } = account;
  if (!shopId) return null;

  const result = {
    _id: shopId
  };

  return xformShopResponse(result);
}
