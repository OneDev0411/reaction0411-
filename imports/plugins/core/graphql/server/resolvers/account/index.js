import Account from "./Account";
import AccountConnection from "./AccountConnection";
import AccountEdge from "./AccountEdge";
import AddAccountAddressBookEntryPayload from "./AddAccountAddressBookEntryPayload";
import Group from "./Group";
import GroupConnection from "./GroupConnection";
import GroupEdge from "./GroupEdge";
import Mutation from "./Mutation";
import Query from "./Query";
import RoleConnection from "./RoleConnection";
import RoleEdge from "./RoleEdge";

export default {
  Account,
  AccountConnection,
  AccountEdge,
  AddAccountAddressBookEntryPayload,
  Group,
  GroupConnection,
  GroupEdge,
  Mutation,
  Query,
  RoleConnection,
  RoleEdge
};
