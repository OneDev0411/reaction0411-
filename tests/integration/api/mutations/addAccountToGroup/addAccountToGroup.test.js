import { encodeAccountOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/account";
import { encodeGroupOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/group";
import { encodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";
import Factory from "/tests/util/factory.js";
import TestApp from "/tests/util/TestApp.js";
import AddAccountToGroupMutation from "./AddAccountToGroupMutation.graphql";

jest.setTimeout(300000);

let accountOpaqueId;
let addAccountToGroup;
let customerGroup;
let customerGroupOpaqueId;
let mockAdminAccount;
let mockAdminAccountWithMissingPermission;
let mockOtherAccount;
let shopId;
let shopManagerGroup;
let shopManagerGroupOpaqueId;
let shopOpaqueId;
let testApp;

beforeAll(async () => {
  testApp = new TestApp();
  await testApp.start();
  shopId = await testApp.insertPrimaryShop();

  mockAdminAccount = Factory.Account.makeOne({
    _id: "mockAdminAccount",
    roles: {
      [shopId]: ["admin", "shopManagerGroupPermission", "someOtherPermission", "customerGroupPermission"]
    },
    shopId
  });
  await testApp.createUserAndAccount(mockAdminAccount);

  mockAdminAccountWithMissingPermission = Factory.Account.makeOne({
    _id: "mockAdminAccountWithMissingPermission",
    roles: {
      [shopId]: ["admin", "someOtherPermission"]
    },
    shopId
  });
  await testApp.createUserAndAccount(mockAdminAccountWithMissingPermission);

  mockOtherAccount = Factory.Account.makeOne({
    _id: "mockOtherAccount",
    groups: [],
    roles: {},
    shopId
  });
  await testApp.createUserAndAccount(mockOtherAccount);

  shopManagerGroup = Factory.Group.makeOne({
    createdBy: null,
    name: "shop manager",
    permissions: ["admin", "shopManagerGroupPermission"],
    slug: "shop manager",
    shopId
  });
  await testApp.collections.Groups.insertOne(shopManagerGroup);

  customerGroup = Factory.Group.makeOne({
    createdBy: null,
    name: "customer",
    permissions: ["customerGroupPermission"],
    slug: "customer",
    shopId
  });
  await testApp.collections.Groups.insertOne(customerGroup);

  accountOpaqueId = encodeAccountOpaqueId(mockOtherAccount._id);
  customerGroupOpaqueId = encodeGroupOpaqueId(customerGroup._id);
  shopOpaqueId = encodeShopOpaqueId(shopId);
  shopManagerGroupOpaqueId = encodeGroupOpaqueId(shopManagerGroup._id);

  addAccountToGroup = testApp.mutate(AddAccountToGroupMutation);
});

afterAll(async () => {
  await testApp.stop();
});

beforeEach(async () => {
  await testApp.collections.Accounts.updateOne({ _id: mockOtherAccount._id }, {
    $set: {
      groups: []
    }
  });

  await testApp.collections.users.updateOne({ _id: mockOtherAccount._id }, {
    $set: {
      roles: {}
    }
  });
});

test("anyone can add account to group if they have ALL the group permissions", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  const result = await addAccountToGroup({ accountId: accountOpaqueId, groupId: shopManagerGroupOpaqueId });

  expect(result.addAccountToGroup.group).toEqual({
    _id: shopManagerGroupOpaqueId,
    createdAt: shopManagerGroup.createdAt.toISOString(),
    createdBy: null,
    description: shopManagerGroup.description,
    name: shopManagerGroup.name,
    permissions: shopManagerGroup.permissions,
    shop: {
      _id: shopOpaqueId
    },
    slug: shopManagerGroup.slug,
    updatedAt: shopManagerGroup.updatedAt.toISOString()
  });

  const account = await testApp.collections.Accounts.findOne({ _id: mockOtherAccount._id });
  expect(account.groups).toEqual([shopManagerGroup._id]);

  const user = await testApp.collections.users.findOne({ _id: mockOtherAccount._id });
  expect(user.roles[shopId]).toEqual(shopManagerGroup.permissions);
});

test("anyone cannot add account to group if they do not have ALL the group permissions", async () => {
  await testApp.setLoggedInUser(mockAdminAccountWithMissingPermission);

  const beforeUser = await testApp.collections.users.findOne({ _id: mockOtherAccount._id });
  expect(beforeUser.roles[shopId]).toBe(undefined);

  try {
    await addAccountToGroup({ accountId: accountOpaqueId, groupId: shopManagerGroupOpaqueId });
  } catch (errors) {
    expect(errors[0]).toMatchSnapshot();
  }

  const account = await testApp.collections.Accounts.findOne({ _id: mockOtherAccount._id });
  expect(account.groups.length).toBe(0);

  const user = await testApp.collections.users.findOne({ _id: mockOtherAccount._id });
  expect(user.roles[shopId]).toBe(undefined);
});

test("permissions from the account's previous group are removed", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await addAccountToGroup({ accountId: accountOpaqueId, groupId: shopManagerGroupOpaqueId });

  let account = await testApp.collections.Accounts.findOne({ _id: mockOtherAccount._id });
  expect(account.groups).toEqual([shopManagerGroup._id]);

  let user = await testApp.collections.users.findOne({ _id: mockOtherAccount._id });
  expect(user.roles[shopId]).toEqual(shopManagerGroup.permissions);

  await addAccountToGroup({ accountId: accountOpaqueId, groupId: customerGroupOpaqueId });

  account = await testApp.collections.Accounts.findOne({ _id: mockOtherAccount._id });
  expect(account.groups).toEqual([customerGroup._id]);

  user = await testApp.collections.users.findOne({ _id: mockOtherAccount._id });
  expect(user.roles[shopId]).toEqual(customerGroup.permissions);
});
