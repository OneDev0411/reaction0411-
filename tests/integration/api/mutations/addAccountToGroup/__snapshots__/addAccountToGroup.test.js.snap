// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anyone cannot add account to group if they do not have ALL the group permissions 1`] = `[GraphQLError: Access Denied]`;
